### PR TITLE
Feature/twilio

### PR DESCRIPTION
**Pull request by major13ua**
Adds basic twilio messaging capabilities.
Messaging with rest-api, messaging via kafka event.

### CODE REVIEW

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class TwilioChannelHandler implements ChannelHandler {
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java:35]: is it possible to extract common properties and logic from TwilioChannelHandler and TelegramChannelHandler to Abstract class?

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
+package com.icthh.xm.tmf.ms.communication.channel.twilio;
+
+import com.icthh.xm.commons.topic.config.MessageListenerContainerBuilder;
+import com.icthh.xm.commons.topic.domain.ConsumerHolder;
+import com.icthh.xm.commons.topic.domain.TopicConfig;
+import com.icthh.xm.tmf.ms.communication.channel.ChannelHandler;
+import com.icthh.xm.tmf.ms.communication.channel.telegram.KafkaToTelegramMessageHandler;
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java:7]: pls remove unused imports

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/MessageHandlerService.java [hunk 3]
```diff
@@ -1,6 +1,7 @@
 package com.icthh.xm.tmf.ms.communication.messaging.handler;
 
 import com.icthh.xm.tmf.ms.communication.domain.MessageType;
+import com.pengrad.telegrambot.model.Message;
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/MessageHandlerService.java:4]: unused import

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TwilioMessageHandler.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    @Override
+    public void handle(CommunicationMessage message) {
+        CommunicationMessageCreate communicationMessageCreate = INSTANCE.messageToMessageCreate(message);
+        this.handle(communicationMessageCreate);
+    }
+
+    @Override
+    public void handle(CommunicationMessageCreate messageCreate) {
+        CommunicationMessage send = twilioService.send(tenantContextHolder.getTenantKey(), messageCreate);
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TwilioMessageHandler.java:29]: variable 'send' is redundant

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.icthh.xm.tmf.ms.communication.messaging.handler.CommunicationMessageMapper.INSTANCE;
+
+@Slf4j
+@Service
+@IgnoreLogginAspect
+@RequiredArgsConstructor
+public class TwilioService  implements MessageService {
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java:42]: Is it possible to extract common logic from TelegramService and TwillioService to abstract class?

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+        if (CollectionUtils.isEmpty(twilioConfigs)) {
+            stopAllTenantConsumers(tenantKey);
+            twilioService.unregisterTwilioSenders(tenantKey);
+            return;
+        }
+
+        //process channels queues
+        processDefaultTwilioConsumer(tenantKey);
+        // todo custom queues for telegram bots will be implemented in next releases
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java:62]: remove Telegram mentions

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    protected AbstractMessageListenerContainer buildListenerContainer(String tenantKey, TopicConfig topicConfig) {
+        KafkaToTwilioMessageHandler messageHandler = new KafkaToTwilioMessageHandler(twilioService);
+        return new MessageListenerContainerBuilder(kafkaProperties, kafkaTemplate)
+            .build(tenantKey, topicConfig, messageHandler);
+    }
+
+    private void stopAllTenantConsumers(String tenantKey) {
+        Map<String, ConsumerHolder> existingConsumers = getTenantConsumers(tenantKey);
+        Collection<ConsumerHolder> holders = existingConsumers.values();
+        withLog(tenantKey, "stopAllTenantTelegramConsumers", () -> {
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java:96]: Telegram?

File: src/test/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandlerUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        ApplicationProperties.Messaging messaging = new ApplicationProperties.Messaging();
+        messaging.setReciveQueueNameTemplate("communication_%s_%s_recive");
+        messaging.setSendQueueNameTemplate("communication_%s_%s_send");
+        when(applicationProperties.getMessaging()).thenReturn(messaging);
+
+        doReturn(container).when(twilioChannelHandler).buildListenerContainer(any(), any());
+    }
+
+    @Test
+    public void testStartingOfNewTelegramBots() {
```
Comment (amedvedchuk) [inline src/test/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandlerUnitTest.java:61]: Telegram?

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+            return;
+        }
+        withLog("twilioSender", () -> startTwilioSender(tenantKey, twilioConfig), twilioConfig.getKey());
+    }
+
+    public void unregisterTwilioSenders(String tenantKey) {
+        twilioClients.remove(tenantKey);
+    }
+
+    protected MessageCreator byNumberOrSid(Receiver receiver, CommunicationMessageCreate message, String serviceId) {
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java:103]: the method is never used

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java (hunk 1 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java:35]: Will be refactored after @sergeysenja1992 MR

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java (hunk 5 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java:42]: Will be refactored after @sergeysenja1992 MR

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TwilioMessageHandler.java (hunk 4 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/TwilioMessageHandler.java:29]: added logs

File: src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/MessageHandlerService.java (hunk 3 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/messaging/handler/MessageHandlerService.java:4]: removed

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java (hunk 6 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java:62]: removed

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java (hunk 5 above)
Comment (sergeysenja1992) [reply src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java:42]: не завязывайся на мой MR
я его сейчас закрою
буду делать по другому

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java (hunk 2 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java:7]: removed

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java (hunk 5 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java:42]: Тогда сделаю рефакторинг MessageService для Twilio/Telegram на следующей итерации.

File: src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java (hunk 9 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/service/TwilioService.java:103]: Нужен для след итерации.

File: src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java (hunk 7 above)
Comment (major13ua) [reply src/main/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandler.java]: changed

File: src/test/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandlerUnitTest.java (hunk 8 above)
Comment (major13ua) [reply src/test/java/com/icthh/xm/tmf/ms/communication/channel/twilio/TwilioChannelHandlerUnitTest.java]: changed

**Review by amedvedchuk — APPROVED** [review]

**Review by ishkurko — APPROVED** [review]

**Pull request merged**